src/trace/trace.cpp

```diff
@@ -845,57 +652,36 @@ void Trace::add_threads(const std::unordered_map<pid_t, std::string>& tid_map)
 {
-    // TODO, should be praradigm_type::measurement_system, but that's a bug in Vampir
-    auto ret = regions_groups_monitoring_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(name),
-        std::forward_as_tuple(group_ref(), intern(name), otf2::common::paradigm_type::user,
-                              otf2::common::group_flag_type::none));
-    return ret.first->second;
+    return registry_.create<otf2::definition::source_code_location>(ByLineInfo(info),
+                                                                    intern(info.file), info.line);
 }
 
-otf2::definition::comm Trace::process_comm(pid_t pid)
+const otf2::definition::region& Trace::intern_region(const LineInfo& info)
 {
-    return process_comms_.at(pid);
-}
-
-otf2::definition::source_code_location Trace::intern_scl(const LineInfo& info)
-{
-    auto ret = source_code_locations_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(info),
-        std::forward_as_tuple(scl_ref(), intern(info.file), info.line));
-    return ret.first->second;
-}
-
-otf2::definition::region Trace::intern_region(const LineInfo& info)
-{
-    auto name_str = intern(info.function);
-    auto ret = regions_line_info_.emplace(
-        std::piecewise_construct, std::forward_as_tuple(info),
-        std::forward_as_tuple(region_ref(), name_str, name_str, name_str,
-                              otf2::common::role_type::function,
-                              otf2::common::paradigm_type::sampling, otf2::common::flags_type::none,
-                              intern(info.file), info.line, 0));
-    return ret.first->second;
+    const auto& name_str = intern(info.function);
+    const auto& region = registry_.create<otf2::definition::region>(
+        ByLineInfo(info), name_str, name_str, name_str, otf2::common::role_type::function,
+        otf2::common::paradigm_type::sampling, otf2::common::flags_type::none, intern(info.file),
+        info.line, 0);
+    if (registry_.has<otf2::definition::regions_group>(ByString(info.dso)))
+    {
+        registry_.get<otf2::definition::regions_group>(ByString(info.dso)).add_member(region);
+    }
+    else
+    {
+        registry_.create<otf2::definition::regions_group>(ByString(info.dso), intern(info.dso),
+                                                          otf2::common::paradigm_type::compiler,
+                                                          otf2::common::group_flag_type::none);
+    }
+    return region;
 }
-
-otf2::definition::string Trace::intern(const std::string& name)
+const otf2::definition::string& Trace::intern(const std::string& name)
 {
     std::lock_guard<std::recursive_mutex> guard(mutex_);
 
-    auto ret = strings_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
-                                std::forward_as_tuple(string_ref(), name));
-    return ret.first->second;
+    return registry_.create<otf2::definition::string>(ByString(name), name);
```

Comment:
This can't work: https://github.com/tud-zih-energy/otf2xx/blob/master/include/otf2xx/registry.hpp#L243

However, seeing this, it might be better to rename/add a `(try_)emplace` method to the registry.